controllers/routing/routing_controller.go

```diff
@@ -84,19 +85,17 @@ func (r *PlatformRoutingController) Reconcile(ctx context.Context, req ctrl.Requ
 
 	var errs []error
 
-	if controllerutil.AddFinalizer(sourceRes, metadata.Finalizers.Routing) {
-		if errUpdate := r.Update(ctx, sourceRes); errUpdate != nil {
-			return ctrl.Result{}, fmt.Errorf("failed adding finalizer: %w", errUpdate)
-		}
+	addFinalizerErr := r.addResourceFinalizer(ctx, req, sourceRes)
+
+	if addFinalizerErr != nil {
+		return ctrl.Result{}, fmt.Errorf("failed adding finalizer with retry: %w", addFinalizerErr)
 	}
 
 	for _, reconciler := range reconcilers {
 		errs = append(errs, reconciler(ctx, sourceRes))
 	}
 
-	if errUpdate := r.Update(ctx, sourceRes); errUpdate != nil {
-		errs = append(errs, errUpdate)
-	}
+	errs = append(errs, r.patchResourceMetadata(ctx, req, sourceRes))
```

Comment:
```suggestion
	errs = append(errs, r.patch(ctx, sourceRes))
```

We can get `NamespacedName` from passed target (`sourceRes`), there is no need to pass `req` around.